### ISSUE TITLE
Term names affect cycle hash when terms are structurally equivalent, leading to equivalent cycles with different hashes, and possibly different cycles with the same hash

### ISSUE THREAD

**Issue by ChrisPenner** [unisonweb/unison#2787]
When reviewing updates to the hashing code I realized that the canonical ordering of component terms is based on the structure of those terms, but importantly, the canonical ordering _ignores self-referential variables_ when it does this.

This works great in most cases, but I wondered, what if two terms were structurally equivalent and differed only in that they had variables referring to different terms within the same cycle?

Here I present 3 cycles, which should all receive the same component hash because they're all equivalent aside from term names;

However, cycle 2 is given a different hash.

As far as I can tell, the reason for this is that `inner1` and `inner2` are structurally equivalent terms, they differ only in that they refer to different terms within their own cycle.

This means that they'll both actually receive the SAME hash when attempting to establish a canonical term ordering, despite being different terms.

As a result, whether inner1 or inner2 comes first in the canonical ordering is dependent on which order they're passed to the hashComponent function, which as far as I can tell, corresponds to the alphabetic ordering of the term names.

So, when I add an `a` in front of `ainnerX2`, it moves it before `innerX1` in the ordering, and thus comes first in the canonical ordering, which results in a different cycle hash for cycle 2.

```
-- Cycle 1
outer1 = 1 + !inner1
outer2 = 2 + !inner2

inner1 _ = 1 + !inner2 + outer1
inner2 _ = 1 + !inner1 + outer2


-- Cycle 2 (These are equivalent to Cycle 1, but get a different component hash!)
outerX1 = 1 + !innerX1
outerX2 = 2 + !ainnerX2

innerX1 _ = 1 + !ainnerX2 + outerX1
ainnerX2 _ = 1 + !innerX1 + outerX2

-- Cycle 3 (These are equivalent to Cycle 1 and are given equal hashes as expected)
outerY1 = 1 + !innerY1
outerY2 = 2 + !innerY2

innerY1 _ = 1 + !innerY2 + outerY1
innerY2 _ = 1 + !innerY1 + outerY2
```

```
.example1> names outer1

  Term
  Hash:   #vilkvev8kh.1c4
  Names:  outer1 outerY1

.example1> names outerX1

  Term
  Hash:   #c5t6op881o.3c4
  Names:  outerX1

```

This is a contrived case of course, but it shows that cycles which are the same can have different hashes, and I suspect (by manipulating this property in the reverse direction) that different cycles can have the _same_ hash (I'll try to prove this with a transcript).

So we may need to come up with a different method of resolving order conflicts like this.

**Comment by aryairani** [comment]
Eek, good find. Sounds like another pre-requisite for #2471 😞 

@pchiusano What do you think about this issue?

> I suspect (by manipulating this property in the reverse direction) that different cycles can have the same hash (I'll try to prove this with a transcript)

Different cycles may have the same hash, but it should be exponentially unlikely, because hashing. Unless there's something about this that sidesteps the crypto?

**Comment by ChrisPenner** [comment]
> Different cycles may have the same hash, but it should be exponentially unlikely, because hashing. Unless there's something about this that sidesteps the crypto?

Yeah, I'm worried it's possible to sidestep the crypto 😬 , it may be possible to exploit the fact that we can control the canonical ordering with something other than the content to manufacture collisions, (only collisions for structurally similar cycles luckily, but this would still be bad). I can think of a few ways it might be possible, I tried once or twice and wasn't able to find an example for it though, but I'm not yet convinced it isn't possible. Regardless, we should probably fix this anyways to prevent the case where re-naming a term causes updates.

I have some plans for how we may be able to fix this AND speed up hashing. Let's chat about it sometime :)

**Comment by pchiusano** [comment]
I've known about this behavior for a while, though I guess I still don't see the problem? If you have a cycle, and one or more elements of the cycle are structurally identical, the order that you pick for those cycle elements is arbitrary. If you do have some idea for a non-arbitrary way to order those elements of the cycle, I'd be interested... though I still don't see the punchline.

I might not be understanding you, but the collision attack does not sound realistic, or it sounds as hard as just coming up with any other hash collision. You could create a nefarious cycle and futz with the variable names to get different permutations and thus different hashes until one collides, but there's no need to jump through those hoops when you could just take the nefarious code and make tweaks to it (like adding an unused `Nat` variable) until you get a hash collision with the code you want to replace. Both these "attacks" are as hard as just breaking SHA3.

**Comment by pchiusano** [comment]
One thing to know is that IIRC after the canonical order is selected, the hash of the cycle is then computed based on that order. So for a cycle with `a`, `b`, and `c`, while the variable reference `a` is initially the same as `b` for purposes of computing the canonical order, after we have a canonical order (say `a`, `c`, `b`), those variable references become like debruijn indexes 0, 1, and 2 and the hash for the cycle will be based on the list of terms `[a,c,b]` in that order. 

So assuming my memory is correct, it shouldn't be possible to get a hash collision unless it is with another cycle which is identical up to reordering of the elements. Which is what I'd expect.

**Comment by ChrisPenner** [comment]
> I've known about this behavior for a while, though I guess I still don't see the problem? If you have a cycle, and one or more elements of the cycle are structurally identical, the order that you pick for those cycle elements is arbitrary.

When I say "structurally identical" I mean that they have the same ABT spine, but not necessarily the same variables, i.e. the first hash (where we treat all variables from within the cycle as equivalent) will be the same. So this means it's possible for two terms which are "structurally equivalent" in this way to still have different behaviours/semantics.

> the order that you pick for those cycle elements is arbitrary

Right, currently the canonical ordering for a cycle in which two or more pieces are "structurally equivalent" is undefined. This is what's causing us to compute two different hashes for the same cycle, I agree that this is going to be an extremely rare edge-case, but definitely still wanted to bring it up, since it's definitely unexpected that changing a term's name changes its hash.

> the collision attack does not sound realistic

I agree that this probably isn't a feasible vector for an _attack_, moreso what I'm saying is that it's still possible for the community to encounter a collision in this way organically given a large enough codebase size. I think this is still pretty unlikely, but still much more likely than an organic hash collision, and it would be a pretty bad experience to have two different cycles competing for the same hash.

> If you do have some idea for a non-arbitrary way to order those elements of the cycle

Yeah, I think I have a way to define an order for structurally equivalent terms, and may be able to bake it in alongside a performance improvement as well (if the perf improvement is worth the extra code it'll take), let's chat in the meeting today 👍🏼

**Comment by calebh** [comment]
Hey all, @ChrisPenner pointed me in the direction of this issue. For the case of a strongly connected graph (which a SCC is) with ordered edges, there is a quadratic time algorithm for canonizing the order of the nodes. In this example I'll be using the following set of functions which form three nodes with identical non-recursive labels, but have different edge structure:

```
h : '()
h _ =
    h()
    foo()

foo : '()
foo _ =
    g()
    h()

g : '()
g _ =
    g()
    h()
```

Here is a picture of the digraph representation of these functions. Note that the edges are ordered based on the location of the function call in a pre-order traversal of that function's AST.

<img width="378" height="354" alt="Image" src="https://github.com/user-attachments/assets/5fcfeb0c-2908-44b6-82cc-42c770c3ace5" />

The first step is to remove all cyclic references and replace them with natural numbers (in pre-order traversal order), then hash the entire SCC at once. This step should be familiar and is well known in the Unison documentation. This becomes the label in the digraph. In this case, the label is:

```
    [0]()
    [1]()
```

Where [0] and [1] are the holes filled with natural numbers.

This is the point where we need to compute a canonical order. Here's how to do it in quadratic time:
1. For each node, *n* walk the graph in a pre-order traversal and construct a tree. If at any point in the walk you hit a previously visited node, insert a dummy leaf instead filled with a natural number instead. For example, let's say node *x* was the n-th node visited in the walk. If you hit node *x* again later in the walk, the dummy leaf will be filled with n-1.
2. Hash this tree using the standard Merkle-style technique.
3. If two trees hash to the same value, then these nodes are truly in the same vertex orbit! In this case the nodes are truly identical.
4. You can now canonically order the nodes and partition the vertices into orbits using the hash of the tree.

Here's some images of the trees produced by this algorithm for the above example. Here I am assuming that the hash of
```
    [0]()
    [1]()
```
is abc12

<img width="400" height="528" alt="Image" src="https://github.com/user-attachments/assets/2f2c4870-e54f-4b69-a50d-3953a9915ea6" />

<img width="442" height="536" alt="Image" src="https://github.com/user-attachments/assets/74ddcf39-a96d-4475-858b-8fc7a3f9ed40" />

<img width="491" height="516" alt="Image" src="https://github.com/user-attachments/assets/6af5b39e-f288-4d1a-8d60-312a51df4df8" />

For each tree it is possible to reconstruct the original digraph - all the information is encoded in the nodes. Note that the order of the children in the tree is important!

As an optimization step, you can skip constructing the tree for a node if its label is unique. The worst case time for this algorithm is quadratic - this happens in a clique graph where every node needs to have it's tree constructed, and to construct the tree you must visit every node.

This algorithm is inspired by a recent paper I've read *Scott : A method for representing graphs as rooted trees for graph canonization*.

For nodes that are truly in the same orbit, I typically take the minimum index of that orbit within the canonical order to refer to that particular node.
